Fix issue #7438 (Crash on Mac Mavericks in traffic light code)

prksingh opened the pull request:
Fixing trafficlightbutton crash https://github.com/adobe/brackets/issues/7438 in Brackets-shell. The crash is UTR so this is not a confirmed fix.

JeffryBooher commented:
I looked at it and the changes look OK by me but @bobeast should have a look if he has time.  Also I need to test it and @redmunds should test it on Yosemite.

JeffryBooher commented:
Fixes https://github.com/adobe/brackets/issues/7438

redmunds commented:
This branch seems to revert to CEF 1547. Is that right?

redmunds commented:
I can build and run Brackets on Yosemite. I did Save As a couple times and did not see any crash.

ingorichter commented on appshell/TrafficLightButton.mm:
```diff
@@ -70,21 +182,21 @@ - (void)setup {
+        self.dirtyPressed = [NSImage imageNamed:[NSString stringWithFormat:@"window-%@-dirty-pressed",buttonName]];
     }
-
+    
```
nit: please remove the trailing spaces

nethip commented:
@prksingh Good catch and change! But this seems to be lot of change. How about just changing to just `active = [NSImage imageNamed:[NSString stringWithFormat:@"window-%@-active",buttonName] retain];` for all the variables instead of these changes. I am afraid too many releases and retains are going to put app to an inconsistent state. And also `NSImage` returned by `[NSImage imageNeeded]` is `autorelelease`ed and so the OS is not going to release it as long as some one is holding on to it's reference, unless someone calls class `[NSAutoReleasePool release]`. Nevertheless, this is a safe change as long as the no of lines of code is put to minimum.

I think the actual culprit here is `[NSImage imageNeeded]` itself. This could be easily substituted with other ways of loading `NSImage` where we are responsible for managing memory. So that way we could just rule out any memory issues, if any.

JeffryBooher commented:
@prksingh also @redmunds ran into an issue where I think you just need to merge master into this branch so it can be tested.  It may have been based on an older branch because I don't see the change  to revert CEF expressly made.

prksingh commented:
@nethip I can remove the setters and move the retain to setup(). My only worry was in case this method gets executed multiple times we may be retaining multiple copies that could cause a leak. It's great if the `autorelease` refcounting is able to handle this. Thanks for the quick review

prksingh commented:
@JeffryBooher @redmunds @ingorichter Thanks for looking at this. I have updated the branch from master and removed the trailing spaces. It have also updated the changes as requested by @nethip.

Please let me know if i missed anything else. Thanks!

redmunds commented:
Looks good on Yosemite.

nethip commented:
@redmunds thanks for testing it out. Is it possible for you to give it a spin on Mavericks, if you have that setup already. If not, could anyone else give it a spin on Mavericks.

peterflynn commented:
Awesome to see a fix for this!  I've tweaked the title a little to be more easy to understand when looking at bug/PR lists.
